### PR TITLE
Use default port 443 in HTTP CONNECT request

### LINKED ISSUE #26227: http_proxy support can issue non-compliant CONNECT requests
### What version of gRPC and what language are you using?
grpcio==1.37.1 via python

### What operating system (Linux, Windows,...) and version?
Linux (but built from source using pip)

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.8 and likely clang

### What did you do?

Following up on https://github.com/googleads/google-ads-python/issues/423 it appears that with a http proxy set, grpc is can issue CONNECT requests that lack a port number.  This was triggered by a change in google-ads in release 10.0.0 that I assume works in the non-proxy case.

### What did you expect to see?

I would expect that either grpc validates that port is present, or grpc adds a default port.

The port is required as part of the "authority form", per

https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6
https://datatracker.ietf.org/doc/html/rfc7230#section-5.3.3

In google-ads==9.0.0, it provides the full form to grpc, and this works.

```
CONNECT googleads.googleapis.com:443 HTTP/1.0
```

### What did you see instead?

```
CONNECT googleads.googleapis.com HTTP/1.0
```

In google-ads==10.0.0, it omits the port, and grpc passes this along to the proxy which then rejects it as invalid.
This shows as "failed to connect to all addresses" with no further information unless you enable `GRPC_TRACE`.

More logs are in the linked google-ads issue.

### PR DESCRIPTION
**Pull request by yashykt**
Fix #26227

### CODE REVIEW

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 1]
```diff
@@ -107,6 +107,17 @@ char* GetHttpProxyServer(const grpc_channel_args* args, char** user_cred) {
   return proxy_name;
 }
 
+// Adds the default port (443) if target does not contain a port.
+std::string MaybeAddDefaultPort(absl::string_view target) {
+  absl::string_view host;
+  absl::string_view port;
+  SplitHostPort(target, &host, &port);
+  if (port.empty()) {
+    return JoinHostPort(host, 443);
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/http_proxy.cc:116]: I'd ideally prefer not to hard-code the default this way, independently of the default in the DNS resolver that specifies the same thing.  But unfortunately, I don't have a better suggestion for how to handle this, so I guess we can live with it.

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 2]
```diff
@@ -176,7 +187,8 @@ class HttpProxyMapper : public ProxyMapperInterface {
     grpc_arg args_to_add[2];
     args_to_add[0] = grpc_channel_arg_string_create(
         const_cast<char*>(GRPC_ARG_HTTP_CONNECT_SERVER),
-        const_cast<char*>(absl::StripPrefix(uri->path(), "/").data()));
+        const_cast<char*>(
+            MaybeAddDefaultPort(absl::StripPrefix(uri->path(), "/")).data()));
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/http_proxy.cc:191]: Please use `c_str()` instead of `data()` here.  (Pre-existing bug.)

**Review by markdroth — APPROVED** [review]

File: src/core/ext/filters/client_channel/http_proxy.cc (hunk 2 above)
Comment (yashykt) [reply src/core/ext/filters/client_channel/http_proxy.cc:191]: starting c++ 11 c_str() and data() do the same thing https://en.cppreference.com/w/cpp/string/basic_string/data

File: src/core/ext/filters/client_channel/http_proxy.cc (hunk 1 above)
Comment (yashykt) [reply src/core/ext/filters/client_channel/http_proxy.cc:116]: I can make it a constant. Something like kDefaultSecurePort that is shared between the resolver and this.

File: src/core/ext/filters/client_channel/http_proxy.cc (hunk 2 above)
Comment (yashykt) [reply src/core/ext/filters/client_channel/http_proxy.cc]: made the change anyway

File: src/core/ext/filters/client_channel/http_proxy.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/http_proxy.cc:116]: That sounds fine.  It's already defined independently in both of the DNS resolver implementations:

https://github.com/grpc/grpc/blob/c97da0fc25b9bd8d1eead7b76d7a8a1b60625956/src/core/ext/filters/client_channel/resolver/dns/native/dns_resolver.cc#L51

https://github.com/grpc/grpc/blob/c97da0fc25b9bd8d1eead7b76d7a8a1b60625956/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc#L63

**Pull request merged**